### DISCUSSION TITLE
Docs enhancement: Apache proxy example

### DISCUSSION THREAD

**Discussion by zv0n** [advplyr/audiobookshelf#126] [Ideas]
I think it would be nice to have an example of an Apache (or NGINX) config to proxy into a running instance of the program. I like to create proxies to programs so I can use https through Apache, usually a simple ProxyPass works but since this project uses websockets it was a bit more complicated.

I managed to get proxy working with the following configuration:
```
ProxyPreserveHost On
ProxyPass / http://127.0.0.1:1337/
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/?(.*) "ws://127.0.0.1:1337/$1" [P,L]
```

I think this could save a lot of people some time if it were in README or docs.

**Comment by jarrodCoombes** [comment]
Is that example for Apache?

If so, any ideas what it should look like for nginx?

I have this for nginx:

```
server
{
        listen 443 ssl;
        server_name <server>.<domain>.<tld>;

        access_log /var/log/nginx/access.log;
        error_log /var/log/nginx/error.log;

        ssl_certificate      /etc/letsencrypt/live/<cert>;
        ssl_certificate_key  /etc/letsencrypt/live/<key>;

        location / {
                     proxy_set_header X-Real-IP         $remote_addr;
                     proxy_set_header X-Forwarded-For   $proxy_add_x_forwarded_for;
                     proxy_set_header X-Forwarded-Proto https;
                     proxy_set_header X-Forwarded-Host  $http_host;
                     proxy_set_header Host              $http_host;
                     proxy_max_temp_file_size           0;
                     proxy_pass                         http://<local_server>;
                     proxy_redirect                     http:// https://;
                   }
}

```

It sort of works, but not entirely. I can't play anything and some of the admin stuff is not working.

**Comment by advplyr** [comment]
Yeah I think that would be helpful. Can you open a pull request putting that in the readme between Install and Run from Source?
The website docs are going to need a big revamp soon.

**Reply by jarrodCoombes** [reply]
I am happy to do this for nginx, but only once I have an actual working example. The one I posted here *does not* work, and I am not sure how to fix it, but I will pound away at it to see if I can get it to work 100% and then add it.

**Reply by advplyr** [reply]
That would be great. You are running into an issue with the web socket, maybe I can point you in the right direction. The project uses socket.io and they have docs about this: https://socket.io/docs/v4/reverse-proxy/

**Reply by jarrodCoombes** [reply]
Submitted the changes to the readme with the nginx stuff in it.

**Comment by jarrodCoombes** [comment]
Got it working, So for anyone with Nginx this is what it should look like

```
server
{

        listen 443 ssl;
        server_name <sub>.<domain>.<tld>;

        access_log /var/log/nginx/access.log;
        error_log /var/log/nginx/error.log;

        ssl_certificate      <certificate>;
        ssl_certificate_key  <key>;

        location / {
                     proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
                     proxy_set_header  X-Forwarded-Proto $scheme;
                     proxy_set_header  Host              $host;
                     proxy_set_header Upgrade            $http_upgrade;
                     proxy_set_header Connection         "upgrade";

                     proxy_http_version                  1.1;

                     proxy_pass                          http://<URL to forward to>;
                     proxy_redirect                      http:// https://;
                   }
}
```

Works 100%. Will see about doing a pull to update the readme as soon as I can.

**Reply by Jon-NJ** [reply]
I would appreciate any clues on getting this to work in nginx with a location different than "/". Typically I have seen that there must be some "url" or "web root" configuration which defaults to "/" but can be changed to something like "/audiobookshelf" and proxy_pass of "http://127.0.0.1:1337/audiobookshelf". I assume there is no feature for this yet so it would also involve software changes. Then there is probably also the question of whether the Android App can work with such a URL.

**Reply by jarrodCoombes** [reply]
I am assuming you tried with:

```
ssl_certificate_key  <key>;

        location /path/to/stuff {
                    ....
                     proxy_pass                          http://<URL to forward to>;
                     ....
                     proxy_pass                          http://127.0.0.1:1337/path/of/audiobooks;
                     }
```

This would mean that `https://<sub>.<domain>.<tld>/path/to/stuff` would be proxied to `http://127.0.0.1:1337/path/of/audiobooks`

I don't think that any modifications to the app or Audio Bookshelf would be all that necessary for this to work as it the functionality built into nginx. Also, as I understand it, `/path/to/stuff`is not a folder on the server and therefore does not have to exist in the file system, instead it is used for host header matching.

**Comment by Jon-NJ** [comment]
Here is what I tried:
```
upstream audiobookshelf {
    server 127.0.0.1:1337;
}

```
```
	location /audiobookshelf/ {
		proxy_set_header        X-Real-IP $remote_addr;
		proxy_set_header        Host $host;
		proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header        X-Forwarded-Proto $scheme;
		proxy_set_header        Upgrade $http_upgrade;
		proxy_set_header        Connection "upgrade";
		proxy_http_version      1.1;
		proxy_redirect          http:// https://;
		proxy_pass              http://audiobookshelf/;
	}

```
It just keeps spinning. Of course, there could be something wrong with this. A lot is just copy/paste. But my understanding was that "https://mysite.home/audiobookshelf/xxx" should then map to "http://127.0.0.1:1337/xxx". I haven't spent much time yet trying to diagnose it. It does work directly with http://myhomeserver:1337" so at least I am in business.

**Reply by jarrodCoombes** [reply]
So what you are saying is that `http://myhomeserver:1337` works fine, and that is the url you can access audiobooks shelf on? And this is running on the same box as the nginx proxy?

I *think* your config would look like this:



```
server
{
        listen 443 ssl;
        server_name <name/IP of the box running nginx>;

        access_log /var/log/nginx/access.log;
        error_log /var/log/nginx/error.log;

        ssl_certificate      <certificate>;
        ssl_certificate_key  <key>;

        location /audiobooks/ {
                     proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
                     proxy_set_header  X-Forwarded-Proto $scheme;
                     proxy_set_header  Host              $host;
                     proxy_set_header Upgrade            $http_upgrade;
                     proxy_set_header Connection         "upgrade";

                     proxy_http_version                  1.1;

                     proxy_pass                          http://myhomeserver:1337;
                     proxy_redirect                      http:// https://;
                   }
}
```

This means that when you browse to `https://<name/IP of the box running nginx>/audiobooks/` you should be proxied to `http://myhomeserver:1337` (this assumes `myhomeserver` resolves to the ip of the box running the Audiobooks server).

In my setup, the nginx proxy is running on a different box from that of my audiobooks server, however the theory is the same.

**Reply by Jon-NJ** [reply]
Sorry if there was any confusion. Both URL's are for the same computer. It was just that one was a direct host name and one was a virtual server with the same IP address configured in /etc/hosts or a local DNS.

**Reply by jarrodCoombes** [reply]
Got it, then you just need to replace `<name/IP of the box running nginx>` with `myhomeserver` and you should be good. You could also use either:

`                     proxy_pass                          http://myhomeserver:1337;`
OR
`                     proxy_pass                          http://127.0.0.1:1337;`

Either should work without issue.

**Comment by Jon-NJ** [comment]
I looked into why it is not working a bit. There is a 301 (permanent redirect) to "/login". Instead I should have "/audiobooks/login". I don't know vue and JS Node is not my forté. But it looks to me like whatever function is called for redirect in client/middleware/authenticated.js should be prepending with my prefix (which the docker setup does not know about) in order for this to work. Maybe this is something I can look into when I find more time. I also see "serverUrl" and "baseUrl" in nuxtconfig.js. So maybe there is already some framework for doing this already. But I am just guessing at the moment.

**Reply by advplyr** [reply]
In nuxt.config.js you can set the base like
```js
router: {
    base: '/audiobooks'
}
```
It may be possible for me to add an environment variable you can set when running docker that would set that base url. I can test that out later today.

**Reply by advplyr** [reply]
I actually just had to solve a problem around setting the base url and using a proxy service with axios. I shared my solution here https://github.com/nuxt/nuxt.js/issues/7028#issuecomment-949721962 but it still may make this more difficult then simply setting an environment variable.

**Comment by Helico-Pter** [comment]
Thought it would be worth mentioning that ABS is fully accessible and works great when proxied using Nginx-Proxy-Manager. https://nginxproxymanager.com/

**Reply by Koliera** [reply]
Hi, any chance I can have a chat about your setup? I have NPM and it did work some time ago, but now the app isn't connecting properly to the server even though I can connect to it on my PC fine.
